// @adonisjs/ace/build/index.js
"use strict";
/*
 * @adonisjs/ace
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
const fancy_logs_1 = require("@poppinss/fancy-logs");
const logger = new fancy_logs_1.Logger();
exports.logger = logger;
var Kernel_1 = require("./src/Kernel");
Object.defineProperty(exports, "Kernel", { enumerable: true, get: function () { return Kernel_1.Kernel; } });
var Manifest_1 = require("./src/Manifest");
Object.defineProperty(exports, "Manifest", { enumerable: true, get: function () { return Manifest_1.Manifest; } });
var args_1 = require("./src/Decorators/args");
Object.defineProperty(exports, "args", { enumerable: true, get: function () { return args_1.args; } });
var flags_1 = require("./src/Decorators/flags");
Object.defineProperty(exports, "flags", { enumerable: true, get: function () { return flags_1.flags; } });
var BaseCommand_1 = require("./src/BaseCommand");
Object.defineProperty(exports, "BaseCommand", { enumerable: true, get: function () { return BaseCommand_1.BaseCommand; } });
var handleError_1 = require("./src/utils/handleError");
Object.defineProperty(exports, "handleError", { enumerable: true, get: function () { return handleError_1.handleError; } });
var listDirectoryFiles_1 = require("./src/utils/listDirectoryFiles");
Object.defineProperty(exports, "listDirectoryFiles", { enumerable: true, get: function () { return listDirectoryFiles_1.listDirectoryFiles; } });
