// @adonisjs/cli/build/src/Services/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const boxen_1 = __importDefault(require("boxen"));
const kleur_1 = require("kleur");
function logInfo(command, message, helpText) {
    command.$info(`${message} ${helpText ? command.colors.dim().yellow(helpText) : ''}`);
}
exports.logInfo = logInfo;
function logTsCompilerError(title, body) {
    console.error(boxen_1.default(`${title}\n${body}`, {
        padding: {
            top: 0,
            bottom: 0,
            left: 1,
            right: 1,
        },
        borderColor: 'red',
    }));
}
exports.logTsCompilerError = logTsCompilerError;
function logPairs(command, textPairs) {
    let message = '';
    textPairs.forEach((pair) => {
        message += `${pair[0]}${command.colors.dim().yellow(pair[1])}`;
    });
    command.$info(message);
}
exports.logPairs = logPairs;
function dumpAsciiLogo() {
    console.log(require('gradient-string').rainbow("    _       _             _         _     \n   / \\   __| | ___  _ __ (_)___    | |___ \n  / _ \\ / _` |/ _ \\| '_ \\| / __|_  | / __|\n / ___ \\ (_| | (_) | | | | \\__ \\ |_| \\__ \\\n/_/   \\_\\__,_|\\___/|_| |_|_|___/\\___/|___/\n"));
}
exports.dumpAsciiLogo = dumpAsciiLogo;
function logCreateAction(message) {
    console.log(` ${kleur_1.green('create')}     ${message}`);
}
exports.logCreateAction = logCreateAction;
function logError(message) {
    console.log(` ${kleur_1.red('error')}      ${message}`);
}
exports.logError = logError;
