// @adonisjs/fold/build/src/Ioc/Tracer.js
"use strict";
/*
 * @adonisjs/fold
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tracer = void 0;
const events_1 = require("events");
/**
 * Tracer is used to emit event from the IoC container
 * at different steps. Read the guides to understand
 * how tracer works
 */
class Tracer extends events_1.EventEmitter {
    constructor() {
        super(...arguments);
        this.namespaces = [];
    }
    in(namespace, cached) {
        const parent = this.namespaces[this.namespaces.length - 1];
        this.emit('use', { namespace, cached, parent });
        this.namespaces.push(namespace);
    }
    out() {
        this.namespaces.pop();
    }
}
exports.Tracer = Tracer;
/**
 * Null object
 */
class NoopTracer extends events_1.EventEmitter {
    in() { }
    out() { }
}
function tracer(enabled) {
    return enabled ? new Tracer() : new NoopTracer();
}
exports.default = tracer;
